toolkit/tools/scheduler/schedulerutils/buildworker.go

```diff
@@ -193,11 +197,44 @@ func buildSRPMFile(agent buildagents.BuildAgent, buildAttempts int, srpmFile, ou
 	)
 
 	logBaseName := filepath.Base(srpmFile) + ".log"
-	err = retry.Run(func() (buildErr error) {
-		builtFiles, logFile, buildErr = agent.BuildPackage(srpmFile, logBaseName, outArch, dependencies)
-		return
-	}, buildAttempts, retryDuration)
-
+	logger.Log.Debugf("osamatest: with check is (%t)", agent.Config().RunCheck)
+	maxAttempts := 3
+	numAttempts := buildAttempts
+	if agent.Config().RunCheck && buildAttempts < maxAttempts {
+		numAttempts = maxAttempts
+	}
+	wg := new(sync.WaitGroup)
```

Comment:
What's the point of using WaitGroups here? It looks like we're not running any additional threads. I'm a bit confused.

Reply:
When I ran the retry.Run part of the code with 3 attempts, it kicks off one build, then starts another one a second later, then a third after that. The same applied when I made this for loop without the waitgroups. With the waitgroups, the second build attempt doesn't start until after the first. I'm not sure if it's due to some parallelization or something of the like.

Reply:
I see. That was actually going to be my next question: how is the updated approach different from what `retry.Run` was doing? It seems like we're still running `agent.BuildPackage` and continue with the code once it exits. Of course, you've also added the log parsing code, but that should have no effect on the parallel vs sequential builds issue. I think more investigation is in order, to understand why the parallel builds were happening. Or why they have now disappeared.

I'd start by checking if `agent.BuildPackage` is exiting before the package actually finished building.

When it comes to WorkGroups, they make no sense if `Add` and `Done` are called form the same thread. In a simples case, the idea is to call `Add` before you start a thread, which will then call `Done` once it has finished. See: [example](https://gobyexample.com/waitgroups). And `Wait` is meant to halt the main thread (the one which started all the other ones) until all started threads call `Done`.

Reply:
To be completely honest, I tried the WaitGroups in the same section of the code because I wasn't sure if that would help with the package build retries being simultaneous, but I wasn't sure what else to try. I'm not sure why it seems to help where retry.Run didn't. I'll look into agent.BuildPackage and see how that function is returning, or if there is some parallelism somewhere that's causing this.

Reply:
So, it appears that the `chrootAgent` that is used for `agent.BuildPackage` is initialized with multithreading, according to `toolkit/scripts/toolchain/container/toolchain_build_in_chroot.sh`. I'm investigating further to see if that's actually leading to `rpmbuild` running in parallel